src/classes/BDI_AdditionalObjectService.cls

```diff
@@ -0,0 +1,449 @@
+    
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of Salesforce.org nor the names of
+      its contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+ 
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2019
+* @group Batch Data Import
+* @group-content ../../ApexDocContent/BatchDataImport.htm
+* @description Implements the ability to import additional objects during batch data import.
+*/
+
+public with sharing class BDI_AdditionalObjectService {
+
+    /*******************************************************************************************************
+    * @description the data import service we are invoked from
+    */
+    private BDI_DataImportService dataImportService;
+    private BDI_FieldMappingCustomMetadata fieldMappingCustomMetadata;
+
+    private Map<Integer,Map<String,Data_Import_Object_Mapping__mdt>> tierToObjMappings = 
+        new Map<Integer,Map<String,Data_Import_Object_Mapping__mdt>>();
+
+    private Map<String,Map<String,Data_Import_Object_Mapping__mdt>> predecessorDevNameToSuccessors = 
+        new Map<String,Map<String,Data_Import_Object_Mapping__mdt>>();
+        
+    private Set<String> importedRecordFieldNames = new Set<String>();
+
+    public Set<String> objMappingDevNames = new Set<String>();
+
+    /*******************************************************************************************************
+    * @description Constructor
+    * @param dataImportService The data import service instance that invokes us
+    */
+    public BDI_AdditionalObjectService(BDI_DataImportService dataImportService) {
+        this.dataImportService = dataImportService;
+        this.fieldMappingCustomMetadata = BDI_FieldMappingCustomMetadata.getInstance();
```

Comment:
Suggestion: get the field Mapper that is a member of the data import service rather than accessing through the singleton